src/System.Private.CoreLib/shared/System/Text/Utf8Utility.Validation.cs

```diff
@@ -0,0 +1,659 @@
+                lastBufferPosProcessed = ref inputBuffer;
+#endif
+
+                // First, check for the common case of all-ASCII bytes.
+
+                if (DWordAllBytesAreAscii(thisDWord))
+                {
+                    // We read an all-ASCII sequence.
+
+                    inputBuffer = ref Unsafe.Add(ref inputBuffer, sizeof(uint));
+
+                    // If we saw a sequence of all ASCII, there's a good chance a significant amount of following data is also ASCII.
+                    // Below is basically unrolled loops with poor man's vectorization.
+
+                    // Below check is "can I read at least five DWORDs from the input stream?"
+                    if ((int)(void*)Unsafe.ByteOffset(ref inputBuffer, ref finalPosWhereCanReadDWordFromInputBuffer) >= 4 * sizeof(uint))
+                    {
+                        // The JIT produces better codegen for aligned reads than it does for
+                        // unaligned reads, and we want the processor to operate at maximum
+                        // efficiency in the loop that follows, so we'll align the references
+                        // now. It's OK to do this without pinning because the GC will never
+                        // move a heap-allocated object in a manner that messes with its
+                        // alignment.
+
+                        {
+                            thisDWord = Unsafe.ReadUnaligned<uint>(ref inputBuffer);
+                            if (!DWordAllBytesAreAscii(thisDWord))
+                            {
+                                goto AfterReadDWordSkipAllBytesAsciiCheck;
+                            }
+                            inputBuffer = ref Unsafe.Add(ref inputBuffer, (IntPtr)GetNumberOfBytesToNextDWordAlignment(ref inputBuffer));
+                        }
+
+                        // At this point, the input buffer offset points to an aligned DWORD.
+                        // We also know that there's enough room to read at least four DWORDs from the stream.
+
+                        ref byte inputBufferFinalRefAtWhichCanSafelyLoop = ref Unsafe.Add(ref finalPosWhereCanReadDWordFromInputBuffer, -3 * sizeof(uint));
+                        do
+                        {
+                            ref uint currentReadPosition = ref Unsafe.As<byte, uint>(ref inputBuffer);
+
+                            if (!DWordAllBytesAreAscii(currentReadPosition | Unsafe.Add(ref currentReadPosition, 1)))
+                            {
+                                goto LoopTerminatedEarlyDueToNonAsciiData;
+                            }
+
+                            if (!DWordAllBytesAreAscii(Unsafe.Add(ref currentReadPosition, 2) | Unsafe.Add(ref currentReadPosition, 3)))
+                            {
+                                inputBuffer = ref Unsafe.Add(ref inputBuffer, 2 * sizeof(uint));
```

Comment:
Why are we adding 8 here (and 16 below)?

Reply:
This is an early exit from the loop. If the loop iterates successfully, we increment the number of bytes read by 16. But if the loop terminates halfway through, we only increment by 8. (There's a goto immediately after this.)

This should be a little clearer in the most recent iteration of the routine.